typescript/src/redemption.ts

```diff
@@ -0,0 +1,299 @@
+    witness: witness,
+    privateKey: decodedWalletPrivateKey.privateKey,
+    compressed: decodedWalletPrivateKey.compressed,
+  })
+
+  const walletAddress = walletKeyRing.getAddress("string")
+
+  // Use the main UTXO as the single transaction input
+  const inputCoins = [
+    bcoin.Coin.fromTX(
+      bcoin.MTX.fromRaw(mainUtxo.transactionHex, "hex"),
+      mainUtxo.outputIndex,
+      -1
+    ),
+  ]
+
+  const transaction = new bcoin.MTX()
+
+  let txFee = 0
+  let totalOutputValue = 0
+
+  // Process the requests
+  for (const request of redemptionRequests) {
+    // Calculate the value of the output by subtracting fee share and treasury
+    // fee for this particular output from the requested amount
+    const outputValue = request.requestedAmount
+      .sub(request.feeShare)
+      .sub(request.treasuryFee)
+
+    // Add the output value to the total output value
+    totalOutputValue += outputValue.toNumber()
+
+    // Add the fee for this particular request to the overall transaction fee
+    txFee += request.feeShare.toNumber()
+
+    // Only allow standard address type to receive the redeemed Bitcoins
+    const address = bcoin.Address.fromString(request.redeemerAddress)
+    if (
+      address.isPubkeyhash() ||
+      address.isWitnessPubkeyhash() ||
+      address.isScripthash() ||
+      address.isWitnessScripthash()
+    ) {
+      transaction.addOutput({
+        script: bcoin.Script.fromAddress(address),
+        value: outputValue.toNumber(),
+      })
+    } else {
+      throw new Error("Redemption address must be P2PKH, P2WPKH, P2SH or P2WSH")
+    }
+  }
+
+  // If there is a change output, add it explicitly to the transaction.
+  // If we did not add this output explicitly, the bcoin library would add it
+  // anyway during funding, but if the value of the change output was very low,
+  // the library would consider it "dust" and add it to the fee rather than
+  // create a new output.
+  const changeOutputValue = mainUtxo.value - totalOutputValue - txFee
+  if (changeOutputValue > 0) {
+    transaction.addOutput({
+      script: bcoin.Script.fromAddress(walletAddress),
+      value: changeOutputValue,
+    })
+  }
+
+  await transaction.fund(inputCoins, {
+    changeAddress: walletAddress,
+    hardFee: txFee,
+    subtractFee: false,
+  })
+
+  transaction.sign(walletKeyRing)
+
+  return {
+    transactionHex: transaction.toRaw().toString("hex"),
+  }
+}
+/**
+ * Builds a length prefixed output script based on the provided address (P2PKH,
+ * P2WPKH, P2SH or P2WSH).
+ * @param address - Bitcoin address from which the output script will be build
+ * @returns The output script as a string.
+ */
+function buildOutputScript(address: string): string {
+  const script = bcoin.Script.fromAddress(address)
+  const scriptAsStr = script.toRaw().toString("hex")
+  const scriptLength = scriptAsStr.length / 2 // get length in bytes
+
+  return "0x" + scriptLength.toString(16) + scriptAsStr
```

Comment:
Let's do it in the same way as we did for tBTC v1 (see https://github.com/keep-network/tbtc.js/blob/9838c06f21082cc050fab5f1304b95d6fa3db6a7/src/Deposit.js#L866):
```
const rawOutputScript = bcoin.Script.fromAddress(address).toRaw()
return `0x${Buffer.concat([Buffer.from([rawOutputScript.length]), rawOutputScript]).toString("hex")}`
```

Reply:
Done in 17157246593ce9f23251e1e09657604775b5def9.